### DISCUSSION TITLE
Trained TTS Vitsmodel Restricted to 11 Seconds of Audio Generation

### DISCUSSION THREAD

**Discussion by shuvohishab** [NVIDIA-NeMo/NeMo#6972] [Q&A] [answered]
By following [this tutorial](https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/Vits_Training.ipynb), I trained a TTS VitsModel.
While performing inference of the trained TTS model to convert text into speech, I observed a limitation where the system generates only up to 11 seconds of audio. So, when dealing with longer texts (e.g., equivalent to 90 seconds of audio), it just produces only first 11 seconds of equivalent audio.
On the other hand, when the text is short and equivalent to 0 to 11 seconds of audio, the system generates audio of the exact length.

Is there a workaround available to generate the complete audio for longer texts?

Note:
* I'm using the latest codebase to implement the TTS model.
* The audio data I am working with varies in length from 0.1 seconds to 63 seconds.

**Comment by shuvohishab** [comment]
@VahidooX could you kindly review my inquiry? Thank you.

**Comment by titu1994** [comment]
@XuesongYang

**Comment by XuesongYang** [comment]
thanks. @treacker. Could you pls have a look?

**Comment by XuesongYang** [comment]
found the same reported issue as well: https://github.com/NVIDIA/NeMo/issues/6998

**Comment by treacker** [comment, marked as answer]
Sorry for late responce. The problem is in `max_len` parameter, which is set by default to 1000 and can't be changed from `convert_text_to_waveform` function. The workaround is either to use `forward` function which is called inside `convert_text_to_waveform` or add option to change `max_len`.  @XuesongYang choose please what is preferred solution

**Reply by shuvohishab** [reply]
Hi @treacker,
Thanks for your kind response and your way of solution. I'll work on it and come back to you soon.

**Reply by shuvohishab** [reply]
@treacker,
It has worked. Thank you once again.

**Reply by PhilipAmadasun** [reply]
@shuvohishab  How did you get it to work? And what is the maximum possible length?  I have to change the source code itself [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/vits.py)? Or I can just change it through some parameter now? Is there a way to increase max_len for other methods, I have the same problem using `from nemo.collections.tts.models.base import SpectrogramGenerator, Vocoder` In the code:
```
import torch
import IPython.display as ipd
from matplotlib.pyplot import imshow
from matplotlib import pyplot as plt
from nemo.collections.tts.models.base import SpectrogramGenerator, Vocoder
import soundfile as sf
import sys

# List pretrained models available in NeMo
print("Spectrogram Models")
print(SpectrogramGenerator.list_available_models())

print()
print("Vocoders")
print(Vocoder.list_available_models())


if torch.cuda.is_available():
  device = "cuda"
else:
  device = "cpu"

# Load our spectrogram and vocoder models onto our CPU
spectrogram_model = SpectrogramGenerator.from_pretrained("tts_en_tacotron2").eval().to(device)
vocoder = Vocoder.from_pretrained("tts_en_hifigan").eval().to(device)

text = "Within the chasmic depths of my mind, vague shapes, sounds and visions fight to be witnessed. A scattering of incoherent memories..... gateways to the blessed palace...... Darkness......Phase equations made flesh.....Darkness.....Waking dreams of a passionless mortal life long though buried....Darkness....Archs of impossible colors claw desperately across the back of my electronic eyes, my machine souls last attempt to ignite conciousness....Darkness....Light, waking neurons fired into life like a birthing galaxy aroused by the sound of plaintiff cries, not human or A.I, not daemon, distant. Ethereal to my spirit."

# Normalize the text and convert it into individual phonemes/tokens.
tokens = spectrogram_model.parse(text, normalize=True)

# Generate spectrogram from text
spectrogram = spectrogram_model.generate_spectrogram(tokens=tokens)

# Invert the spectrogram into audio samples
audio = vocoder.convert_spectrogram_to_audio(spec=spectrogram)

# Convert output from pytorch tensor to numpy array
spectrogram = spectrogram.cpu().detach().numpy()[0]
audio = audio.cpu().detach().numpy()[0]
audio_file = "output_audio.wav"
sf.write(audio_file, audio, samplerate=22050)
print(f"Audio saved to {audio_file}")

print(f'"{text}"\n')

ipd.Audio(audio, rate=22050)
```

**Reply by shuvohishab** [reply]
Hi @PhilipAmadasun,
[_I was experimenting with Vits, you might find your corresponding model script there according to your model_]
That's a long time ago, so I can't remember the full picture now, sorry.[_Also, it's really disappointing that they haven't made this parameter flexible_.]

But there would be following way around you can try, and as long as I can remember, this was the solution:
Update the `max_len` value here: 
1. https://github.com/NVIDIA/NeMo/blob/1882d0d0025832a530063e5b699347253b2ab768/nemo/collections/tts/models/vits.py#L201
2. https://github.com/NVIDIA/NeMo/blob/1882d0d0025832a530063e5b699347253b2ab768/nemo/collections/tts/models/vits.py#L311

I used **max_len=50000** here.
After updating these 2 lines, you'll need to reinstall the NeMo repo[by using `pip install .` from the repo's root directory].
Please inform me if I was able to assist you. Thank you.

**Reply by PhilipAmadasun** [reply]
@shuvohishab  I am now getting some erros but before I even proceed, is the Nemo TextToSpeech even good for Real Time use cases?

**Reply by shuvohishab** [reply]
@PhilipAmadasun,
TBH yes. I mean, from our team's experiments, it came as the best to use in REAL time cases. Execution time was so little comparing to others. You can use RIVA server to deploy your TTS model to infer your model faster.

**Reply by PhilipAmadasun** [reply]
@shuvohishab My local machine has two A6000s and I don't get good inference. It takes like almost 20 seconds or something to generate audio for short text. Maybe I'm not using any of the accelerator parameters. This is the code here. Also this code only produces 11 seconds of Audio. 
```
import torch
import IPython.display as ipd
from matplotlib.pyplot import imshow
from matplotlib import pyplot as plt
from nemo.collections.tts.models.base import SpectrogramGenerator, Vocoder
import soundfile as sf
import sys

# List pretrained models available in NeMo
print("Spectrogram Models")
print(SpectrogramGenerator.list_available_models())

print()
print("Vocoders")
print(Vocoder.list_available_models())


if torch.cuda.is_available():
  device = "cuda"
else:
  device = "cpu"

# Load our spectrogram and vocoder models onto our CPU
spectrogram_model = SpectrogramGenerator.from_pretrained("tts_en_tacotron2").eval().to(device)
vocoder = Vocoder.from_pretrained("tts_en_hifigan").eval().to(device)

text = "Within the chasmic depths of my mind, vague shapes, sounds and visions fight to be witnessed. A scattering of incoherent memories..... gateways to the blessed palace...... Darkness......Phase equations made flesh.....Darkness.....Waking dreams of a passionless mortal life long though buried....Darkness....Archs of impossible colors claw desperately across the back of my electronic eyes, my machine souls last attempt to ignite conciousness....Darkness....Light, waking neurons fired into life like a birthing galaxy aroused by the sound of plaintiff cries, not human or A.I, not daemon, distant. Ethereal to my spirit."

# Normalize the text and convert it into individual phonemes/tokens.
tokens = spectrogram_model.parse(text, normalize=True)

# Generate spectrogram from text
spectrogram = spectrogram_model.generate_spectrogram(tokens=tokens)

# Invert the spectrogram into audio samples
audio = vocoder.convert_spectrogram_to_audio(spec=spectrogram)

# Convert output from pytorch tensor to numpy array
spectrogram = spectrogram.cpu().detach().numpy()[0]
audio = audio.cpu().detach().numpy()[0]
audio_file = "output_audio.wav"
sf.write(audio_file, audio, samplerate=22050)
print(f"Audio saved to {audio_file}")

print(f'"{text}"\n')

ipd.Audio(audio, rate=22050)
```

**Reply by shuvohishab** [reply]
@PhilipAmadasun,
You can try monitoring your GPU usages, whether it loads any memory or not.
By the way, using Vits with CPU the time was longer. But not more than 4~5 seconds.
But, when I used GPU(1xT4) for inference, the execution time was always less then 1 second whatever the length of the audio was.

**Reply by PhilipAmadasun** [reply]
@shuvohishab  You're right! It's not even using my GPUs at all.

**Reply by PhilipAmadasun** [reply]
@shuvohishab As I mentioned earlier I was getting errors with Vit after I changed the max_len, I'll keep working and provide updates.

**Reply by shuvohishab** [reply]
Sure @PhilipAmadasun,
We've had an informative conversation. Hopefully you'll come with positive news soon, waiting eagerly for that. Wishing you luck.

**Reply by PhilipAmadasun** [reply]
@shuvohishab  Yeah after changing the max_len on both lines I get:
```File "/home/philip/miniconda3/envs/fst/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'nemo.collections.asr.parts.mixins.transcription.InternalTranscribeConfig'> for field _internal is not allowed: use default_factory
``` 
When I try to import VitsModel: `from nemo.collections.tts.models import VitsModel`. Its some dataclass error thing. Also I'm not particularly sure which shell in the [tutorial](https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/Vits_Training.ipynb) is actually producing the audio?

**Reply by PhilipAmadasun** [reply]
@shuvohishab Maybe you have some example unit test script I can look at?

**Reply by shuvohishab** [reply]
```
from nemo.collections.tts.models import VitsModel
model = VitsModel.load_from_checkpoint(checkpoint_path=<your finetuned checkpoint path>).cpu().eval()

text_raw = "your text"
tokens = model.parse(text_raw)
audio_pred = model.convert_text_to_waveform(tokens=tokens).cpu().detach().numpy()

import IPython.display as ipd
ipd.Audio(audio_pred, rate=22050)
```
I used this.

**Reply by PhilipAmadasun** [reply]
@shuvohishab Doesn't that use cpu? And also where is the audio gonna get stored? I want a .wav file out of this.

**Reply by shuvohishab** [reply]
Change cpu to `cuda` if you want to infer in GPU.
At first generate the audio, then you can save the audio with `audio_pred `.

**Reply by PhilipAmadasun** [reply]
@shuvohishab 
```
audio_pred = model.convert_text_to_waveform(tokens=tokens).cuda().detach().numpy()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

**Reply by shuvohishab** [reply]
Kindly follow the instruction provided in error log.

**Reply by PhilipAmadasun** [reply]
@shuvohishab  Never mind, forget my last message. I did have another question about the NeMo installation. When I changed the max lengths, did I need to reinstall nemo all over again? I pip installed the previous time.
